hudi-common/src/main/java/org/apache/hudi/metadata/MetadataPartitionType.java

```diff
@@ -18,30 +18,52 @@
 
 package org.apache.hudi.metadata;
 
+import org.apache.hudi.common.config.HoodieMetadataConfig;
+import org.apache.hudi.common.table.HoodieTableMetaClient;
+
+import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.Collections;
 import java.util.List;
 import java.util.Set;
+import java.util.function.BiPredicate;
+import java.util.function.Predicate;
 import java.util.stream.Collectors;
 
 /**
  * Partition types for metadata table.
  */
 public enum MetadataPartitionType {
-  FILES(HoodieTableMetadataUtil.PARTITION_NAME_FILES, "files-"),
-  COLUMN_STATS(HoodieTableMetadataUtil.PARTITION_NAME_COLUMN_STATS, "col-stats-"),
-  BLOOM_FILTERS(HoodieTableMetadataUtil.PARTITION_NAME_BLOOM_FILTERS, "bloom-filters-"),
-  RECORD_INDEX(HoodieTableMetadataUtil.PARTITION_NAME_RECORD_INDEX, "record-index-"),
-  FUNCTIONAL_INDEX(HoodieTableMetadataUtil.PARTITION_NAME_FUNCTIONAL_INDEX_PREFIX, "func-index-");
+  FILES(HoodieTableMetadataUtil.PARTITION_NAME_FILES, "files-",
+      HoodieMetadataConfig::enabled,
+      (metaClient, partitionType) -> metaClient.getTableConfig().isMetadataPartitionAvailable(partitionType)),
+  COLUMN_STATS(HoodieTableMetadataUtil.PARTITION_NAME_COLUMN_STATS, "col-stats-",
+      HoodieMetadataConfig::isColumnStatsIndexEnabled,
+      (metaClient, partitionType) -> metaClient.getTableConfig().isMetadataPartitionAvailable(partitionType)),
+  BLOOM_FILTERS(HoodieTableMetadataUtil.PARTITION_NAME_BLOOM_FILTERS, "bloom-filters-",
+      HoodieMetadataConfig::isBloomFilterIndexEnabled,
+      (metaClient, partitionType) -> metaClient.getTableConfig().isMetadataPartitionAvailable(partitionType)),
+  RECORD_INDEX(HoodieTableMetadataUtil.PARTITION_NAME_RECORD_INDEX, "record-index-",
+      HoodieMetadataConfig::isRecordIndexEnabled,
+      (metaClient, partitionType) -> metaClient.getTableConfig().isMetadataPartitionAvailable(partitionType)),
+  FUNCTIONAL_INDEX(HoodieTableMetadataUtil.PARTITION_NAME_FUNCTIONAL_INDEX_PREFIX, "func-index-",
+      metadataConfig -> false, // no config for functional index, it is created using sql
+      (metaClient, partitionType) -> metaClient.getFunctionalIndexMetadata().isPresent());
 
   // Partition path in metadata table.
   private final String partitionPath;
   // FileId prefix used for all file groups in this partition.
   private final String fileIdPrefix;
+  private final Predicate<HoodieMetadataConfig> isMetadataPartitionEnabled;
+  private final BiPredicate<HoodieTableMetaClient, MetadataPartitionType> isMetadataPartitionAvailable;
 
-  MetadataPartitionType(final String partitionPath, final String fileIdPrefix) {
+  MetadataPartitionType(final String partitionPath, final String fileIdPrefix,
+                        Predicate<HoodieMetadataConfig> isMetadataPartitionEnabled,
+                        BiPredicate<HoodieTableMetaClient, MetadataPartitionType> isMetadataPartitionAvailable) {
```

Comment:
Can we use overridden methods to define custom behavior in the Enum instead of the functions? e.g.,
```
public enum MetadataPartitionType {
  FILES(HoodieTableMetadataUtil.PARTITION_NAME_FILES, "files-") {
    @Override
    public String isMetadataPartitionEnabled(..) {
        return config.enabled();
    }

    ...
  },
  ...;

  // Abstract method to be overridden by each constant
  public abstract boolean isMetadataPartitionEnabled(..);

  public abstract boolean isMetadataPartitionAvailable(..);
}
```